File: EmployeeScheduler.java
import java.util.*;

public class EmployeeScheduler {
    static String[] DAYS = {"Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"};
    static String[] SHIFTS = {"Morning", "Afternoon", "Evening"};
    static Scanner sc = new Scanner(System.in);

    public static void main(String[] args) {
        Map<String, Employee> employees = new LinkedHashMap<>();

        int numEmployees;
        while (true) {
            System.out.print("Enter number of employees: ");
            try {
                numEmployees = Integer.parseInt(sc.nextLine().trim());
                if (numEmployees < 1) {
                    System.out.println("Number must be greater than 0.");
                    continue;
                }
                if (numEmployees < 9) {
                    System.out.print("Not enough employees to cover all shifts for 7 days. Do you want to continue? (Yes/No): ");
                    String choice = sc.nextLine().trim().toLowerCase();
                    if (choice.equals("no")) continue;
                }
                break;
            } catch (Exception e) {
                System.out.println("Please enter a valid number.");
            }
        }

        // Input employee info
        for (int i = 0; i < numEmployees; i++) {
            String name;
            do {
                System.out.print("Enter employee " + (i + 1) + " name: ");
                name = sc.nextLine().trim();
                if (name.isEmpty()) System.out.println("Employee name cannot be blank.");
            } while (name.isEmpty());

            Map<String, String> prefs = new HashMap<>();
            System.out.println("Enter " + name + "'s preferred shifts (Morning/Afternoon/Evening):");
            for (String day : DAYS) {
                System.out.print("  " + day + ": ");
                String pref = sc.nextLine().trim();
                if (!(pref.equalsIgnoreCase("Morning") || pref.equalsIgnoreCase("Afternoon") || pref.equalsIgnoreCase("Evening"))) {
                    System.out.println("Invalid input. Assigning random shift.");
                    pref = SHIFTS[new Random().nextInt(SHIFTS.length)];
                }
                prefs.put(day, capitalize(pref));
            }
            employees.put(name, new Employee(name, prefs));
        }

        Map<String, Map<String, List<String>>> schedule = new LinkedHashMap<>();
        for (String day : DAYS) {
            Map<String, List<String>> shifts = new LinkedHashMap<>();
            for (String shift : SHIFTS) {
                shifts.put(shift, new ArrayList<>());
            }
            schedule.put(day, shifts);
        }

        // Assign shifts
        for (String day : DAYS) {
            for (Employee emp : employees.values()) {
                if (emp.assignedDays >= 5) continue;
                String pref = emp.preferences.get(day);
                if (schedule.get(day).get(pref).size() < 2 && !schedule.get(day).get(pref).contains(emp.name)) {
                    schedule.get(day).get(pref).add(emp.name);
                    emp.assignedDays++;
                } else {
                    boolean assigned = false;
                    for (String alt : SHIFTS) {
                        if (schedule.get(day).get(alt).size() < 2 && !schedule.get(day).get(alt).contains(emp.name)) {
                            schedule.get(day).get(alt).add(emp.name);
                            emp.assignedDays++;
                            assigned = true;
                            break;
                        }
                    }
                    if (!assigned) {
                        int nextIdx = (Arrays.asList(DAYS).indexOf(day) + 1) % DAYS.length;
                        String nextDay = DAYS[nextIdx];
                        for (String alt : SHIFTS) {
                            if (schedule.get(nextDay).get(alt).size() < 2 && !schedule.get(nextDay).get(alt).contains(emp.name) && emp.assignedDays < 5) {
                                schedule.get(nextDay).get(alt).add(emp.name);
                                emp.assignedDays++;
                                break;
                            }
                        }
                    }
                }
            }
        }

        // Fill missing slots with unique employees
        for (String day : DAYS) {
            for (String shift : SHIFTS) {
                while (schedule.get(day).get(shift).size() < 2) {
                    List<Employee> available = new ArrayList<>();
                    for (Employee e : employees.values()) {
                        if (e.assignedDays < 5 && !schedule.get(day).get(shift).contains(e.name)) available.add(e);
                    }
                    if (available.isEmpty()) break;
                    Employee chosen = available.get(new Random().nextInt(available.size()));
                    schedule.get(day).get(shift).add(chosen.name);
                    chosen.assignedDays++;
                }
            }
        }

        // Print final schedule: days as rows, shifts as columns
        System.out.println("\nFinal Weekly Schedule:\n");
        System.out.printf("%-12s%-25s%-25s%-25s%n", "Day", "Morning", "Afternoon", "Evening");
        for (String day : DAYS) {
            System.out.printf("%-12s", day);
            for (String shift : SHIFTS) {
                String val = String.join(", ", schedule.get(day).get(shift));
                System.out.printf("%-25s", val);
            }
            System.out.println();
        }
    }

    static class Employee {
        String name;
        Map<String, String> preferences;
        int assignedDays = 0;

        Employee(String n, Map<String, String> p) {
            name = n;
            preferences = p;
        }
    }

    static String capitalize(String s) {
        if (s.isEmpty()) return s;
        return s.substring(0, 1).toUpperCase() + s.substring(1).toLowerCase();
    }
}

File: Employee_Scheduler.py
import random

DAYS = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]
SHIFTS = ["Morning", "Afternoon", "Evening"]

# Get number of employees
while True:
    try:
        num_employees = int(input("Enter number of employees: "))
        if num_employees < 1:
            print("Number must be greater than 0.")
            continue
        if num_employees < 9:
            choice = input("Not enough employees to cover all shifts for 7 days. Do you want to continue? (Yes/No): ").strip().lower()
            if choice == "no":
                continue
        break
    except ValueError:
        print("Please enter a valid number.")

# Collect employee info
employees = {}
for i in range(num_employees):
    while True:
        name = input(f"Enter employee {i+1} name: ").strip()
        if name:
            break
        print("Employee name cannot be blank.")
    preferences = {}
    print(f"Enter {name}'s preferred shifts for each day (Morning/Afternoon/Evening):")
    for day in DAYS:
        pref = input(f"  {day}: ").strip().capitalize()
        if pref not in SHIFTS:
            print("Invalid input. Assigning random shift.")
            pref = random.choice(SHIFTS)
        preferences[day] = pref
    employees[name] = {"preferences": preferences, "assigned_days": 0}

# Initialize schedule
schedule = {day: {shift: [] for shift in SHIFTS} for day in DAYS}

# Assign shifts
for day in DAYS:
    for emp, data in employees.items():
        if data["assigned_days"] >= 5:
            continue
        pref = data["preferences"][day]
        if len(schedule[day][pref]) < 2 and emp not in schedule[day][pref]:
            schedule[day][pref].append(emp)
            data["assigned_days"] += 1
        else:
            # Try another shift same day
            assigned = False
            for alt in SHIFTS:
                if len(schedule[day][alt]) < 2 and emp not in schedule[day][alt]:
                    schedule[day][alt].append(emp)
                    data["assigned_days"] += 1
                    assigned = True
                    break
            # Try next day if not possible today
            if not assigned:
                next_day = DAYS[(DAYS.index(day) + 1) % len(DAYS)]
                for alt in SHIFTS:
                    if len(schedule[next_day][alt]) < 2 and emp not in schedule[next_day][alt]:
                        schedule[next_day][alt].append(emp)
                        data["assigned_days"] += 1
                        break

# Fill missing slots ensuring unique employees
for day in DAYS:
    for shift in SHIFTS:
        while len(schedule[day][shift]) < 2:
            available = [e for e, d in employees.items() if d["assigned_days"] < 5 and e not in schedule[day][shift]]
            if not available:
                break
            chosen = random.choice(available)
            schedule[day][shift].append(chosen)
            employees[chosen]["assigned_days"] += 1

# Print final schedule: Days as rows, Shifts as columns
print("\nFinal Weekly Schedule:\n")
header = f"{'Day':<12}{'Morning':<25}{'Afternoon':<25}{'Evening':<25}"
print(header)
print("-" * len(header))

for day in DAYS:
    row = f"{day:<12}"
    for shift in SHIFTS:
        val = ", ".join(schedule[day][shift])
        row += f"{val:<25}"
    print(row)
